src/Playwright.Tests/BrowserContextBaseUrlTests.cs

```diff
@@ -0,0 +1,135 @@
+ * to use, copy, modify, merge, publish, distribute, sublicense, and / or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+using System;
+using System.Text;
+using System.Text.Unicode;
+using System.Threading.Tasks;
+using Microsoft.Playwright.NUnit;
+using NUnit.Framework;
+
+namespace Microsoft.Playwright.Tests
+{
+    [Parallelizable(ParallelScope.Self)]
+    public class BrowserContextBaseUrlTests : BrowserTestEx
+    {
+        [PlaywrightTest("browsercontext-base-url.spec.ts", "should construct a new URL when a baseURL in browser.newContext is passed to page.goto")]
+        [Test, Timeout(TestConstants.DefaultTestTimeout)]
+        public async Task ShouldConstructANewURLInBrowserNewContextWhenGoto()
+        {
+            await using var context = await Browser.NewContextAsync(new() { BaseURL = Server.Prefix });
+            var page = await context.NewPageAsync();
+            var response = await page.GotoAsync("/empty.html");
+            Assert.AreEqual(Server.EmptyPage, response.Url);
+        }
+
+        [PlaywrightTest("browsercontext-base-url.spec.ts", "should construct a new URL when a baseURL in browser.newPage is passed to page.goto")]
+        [Test, Timeout(TestConstants.DefaultTestTimeout)]
+        public async Task ShouldConstructANewURLInBrowserNewPageWhenGoto()
+        {
+            var page = await Browser.NewPageAsync(new() { BaseURL = Server.Prefix });
+            var response = await page.GotoAsync("/empty.html");
+            Assert.AreEqual(Server.EmptyPage, response.Url);
+            await page.CloseAsync();
+        }
+
+        [PlaywrightTest("browsercontext-base-url.spec.ts", "should construct the URLs correctly when a baseURL without a trailing slash in browser.newPage is passed to page.goto")]
+        [Test, Timeout(TestConstants.DefaultTestTimeout)]
+        public async Task ShouldConstructTheURLsCorrectlyWihoutTrailingSlash()
+        {
+            var serverUri = new Uri(Server.Prefix);
+            var page = await Browser.NewPageAsync(new() { BaseURL = new Uri(serverUri, "/url-construction").ToString() });
+
+            Assert.AreEqual(new Uri(serverUri, "/mypage.html"), (await page.GotoAsync("mypage.html")).Url);
```

Comment:
Does page.Url return a string or a URL object? If it's a URL object, then we need to fix it in the docs. Otherwise I would change the assertion as suggested (and below):

```suggestion
            Assert.AreEqual(Server.Prefix + "/mypage.html", (await page.GotoAsync("mypage.html")).Url);
```

Reply:
Would be great if you could align this with upstream, then its easier to follow instead of doing URI magic.